### PR TITLE
Migrate `libs-through-symlinks` and `translation` run-make tests to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Everything here works locally... except line 57 (the test works if it is commented out):

```rs
    rustc()
        .arg("-Ztranslate-lang=zh-CN")
        .input("test.rs")
        .sysroot("fakeroot")
        .run_fail()
        .assert_stderr_contains("this is a test message");
```

This will therefore initially fail CI.

try-job: aarch64-gnu
try-job: aarch64-apple
try-job: x86_64-msvc
try-job: i686-mingw
try-job: test-various
try-job: armhf-gnu

### CODE REVIEW

File: tests/run-make/libs-through-symlinks/rmake.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+// checks that compilation succeeds through use of the symlink.
+// See https://github.com/rust-lang/rust/issues/13890
+
+use run_make_support::{cwd, path, rfs, rust_lib_name, rustc};
+
+fn main() {
+    rfs::create_dir("outdir");
+    rustc().input("foo.rs").output(path("outdir").join(rust_lib_name("foo"))).run();
+    rfs::create_symlink(path("outdir").join(rust_lib_name("foo")), rust_lib_name("foo"));
+    rustc().env("RUSTC_LOG", "rustc_metadata::loader").input("bar.rs").run();
```
Comment (jieyouxu) [inline tests/run-make/libs-through-symlinks/rmake.rs:13]: Remark: this RUSTC_LOG is used for debugging, I guess it's fine to keep it in.

File: tests/run-make/libs-through-symlinks/rmake.rs
Comment (jieyouxu) [inline tests/run-make/libs-through-symlinks/rmake.rs:1]: Problem: this needs a `//@ needs-symlink` directive for Windows.

File: tests/run-make/translation/rmake.rs
Comment (jieyouxu) [inline tests/run-make/translation/rmake.rs:1]: Suggestion: can we please group the test cases alongside their comments, like the original Makefile? E.g.

```rs
fn test_builtin_fallback_bundle() {
    // TODO
}
```
Dumping them together is a bit hard to follow.

File: tests/run-make/translation/rmake.rs [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+// of each test case.
+// See https://github.com/rust-lang/rust/pull/95512
+
+//@ needs-symlink
+
+use run_make_support::{path, rfs, rustc};
+
+fn main() {
+    // Check that the test works normally, using the built-in fallback bundle.
+    rustc().input("test.rs").run_fail().assert_stderr_contains("struct literal body without path");
```
Comment (jieyouxu) [inline tests/run-make/translation/rmake.rs:12]: Suggestion: the `RUSTC_TRANSLATION_NO_DEBUG_ASSERT=1` env var might be load-bearing, see https://github.com/rust-lang/rust/blob/69e36d65f93f81f4cbd8a3191c33be3fc201f095/compiler/rustc_errors/src/translation.rs#L109-L118

We probably want to have a `Rustc` factory fn that presets that env var:

```rs
fn rustc_no_translation_assert() -> Rustc {
    rustc().env("RUSTC_TRANSLATION_NO_DEBUG_ASSERT", "1")
}
```

File: tests/run-make/translation/rmake.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        .sysroot("fakeroot")
+        .run_fail()
+        .assert_stderr_contains("this is a test message");
+
+    // Check that the compiler errors out when the sysroot requested cannot be
+    // found. This test might start failing if there actually exists a Klingon
+    // translation of rustc's error messages.
+    rustc()
+        .arg("-Ztranslate-lang=tlh")
+        // .input("test.rs")
```
Comment (jieyouxu) [inline tests/run-make/translation/rmake.rs:69]: Remark: commented out

File: tests/run-make/translation/rmake.rs (hunk 2 above)
Comment (jieyouxu) [reply tests/run-make/translation/rmake.rs:12]: @rustbot author

Comment (Oneirical) [reply tests/run-make/translation/rmake.rs:12]: `rustc()` returns a temporary `&mut Rustc` which cannot be returned at first glance (unless there is some lifetime magic to pull it off).

I added the environment variable to the rustc calls in the test manually, though in my local tests it seemed to have no effect. Then again, this might be due to me using stage 0 compilation.

File: tests/run-make/translation/rmake.rs (hunk 3 above)
Comment (Oneirical) [reply tests/run-make/translation/rmake.rs:69]: Original test:

```make
sysroot-missing:
	$(RUSTC) $< -Ztranslate-lang=tlh 2>&1 | $(CGREP) "missing locale directory"
```

Note how the ususal `test.rs` next to the section header `sysroot-missing:` is absent. I have no idea what the `$<` is slotting in.

File: tests/run-make/translation/rmake.rs (hunk 2 above)
Comment (jieyouxu) [reply tests/run-make/translation/rmake.rs:12]: Oh right, my bad. Hm.

File: tests/run-make/translation/Makefile [hunk 4]
```diff
@@ -1,78 +0,0 @@
...
-	mkdir $(FAKEROOT)/lib/rustlib
-	ln -s $(SYSROOT)/lib/rustlib/* $(FAKEROOT)/lib/rustlib
-	rm -f $(FAKEROOT)/lib/rustlib/src
-	mkdir $(FAKEROOT)/lib/rustlib/src
-	ln -s $(SYSROOT)/lib/rustlib/src/* $(FAKEROOT)/lib/rustlib/src
-	# When download-rustc is enabled, `$(SYSROOT)` will have a share directory. Delete the link to it.
-	rm -f $(FAKEROOT)/share
-	mkdir -p $(FAKEROOT)/share/locale/zh-CN/
-	ln -s $(CURDIR)/working.ftl $(FAKEROOT)/share/locale/zh-CN/basic-translation.ftl
-	$(RUSTC) $< --sysroot $(FAKEROOT) -Ztranslate-lang=zh-CN 2>&1 | $(CGREP) "this is a test message"
```
Comment (jieyouxu) [inline tests/run-make/translation/Makefile:53]: Remark (to myself): $< ?

**Comment by jieyouxu** [comment]
The symlink trouble I observed previously from compiletest and bootstrap likely came from somehow under non-admin mode artifacts can be produced under admin ownership, which doesn't block this PR now that run_make_support symlink handling is rectified.

**Comment by jieyouxu** [comment]
libs-through-symlink extracted to #134829.

**Comment by jieyouxu** [comment]
Superseded by https://github.com/rust-lang/rust/pull/135818 with attribution. Thanks for the PR!
